### ISSUE TITLE
deploy java functions with error occured "AWS::Lambda::Function UPDATE_FAILED Resource handler returned message: 'Uploaded file must be a non-empty zip'"

### ISSUE THREAD

**Issue by ljd1991** [sst/sst#3274]
I use SST to manage infrastructures, and use Quarkus, which is a Java Application Framework, to develop lambda functions.

By reading docs and guides, I found some feature of Quarkus and SST

- In Quarkus, running `gradle build`command will create a `function.zip` file located in `build` directory, and it is used as lambda function package.

- In SST, according to [Configuring Java runtime](https://docs.sst.dev/constructs/Function#configuring-java-runtime) docs, we should place our Lambda function's Code in `build/distributions`, and SST uploads the `build/distributions/*.zip` as the Lambda function's code.

Here is what I do to make Quarkus work with SST

- In `build.gradle` script, I add a task and register to build flow like below

```groovy
tasks.register('copy', Copy) {
    from "build/function.zip"
    into "build/distributions"
}

quarkusBuild.finalizedBy('copy')
```

This build configuration simply copy `function.zip` generated by Quarkus into `build/distributions`

On `sst dev`, it works fine, all Lambda functions are executed properly.

However,  on `sst deploy`, I got error message like below

```
SST v2.24.14

➜  App:     ****
   Stage:   ****
   Region:  ****
   Account: ****

✔  Building...

AppStack PUBLISH_ASSETS_COMPLETE
|  AppStack FunctionSourcemapUploaderPolicy AWS::IAM::Policy CREATE_COMPLETE
|  AppStack appApi/Lambda_GET_--test/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--test1/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_POST_--app/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--app-list/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--app-list AWS::Lambda::Function UPDATE_FAILED Resource handler returned message: "Uploaded file must be a non-empty zip (Service: Lambda, Status Code: 400, Request ID: 75689c02-22a8-4d51-9cc3-1664bfda0a5e)" (RequestToken: af67eb59-9b37-0c17-8f5d-b95a811ec297, HandlerErrorCode: InvalidRequest)
|  AppStack appApi/Lambda_POST_--app AWS::Lambda::Function UPDATE_FAILED Resource update cancelled
|  AppStack appApi/Lambda_GET_--test1 AWS::Lambda::Function UPDATE_FAILED Resource update cancelled
|  AppStack appApi/Lambda_GET_--test AWS::Lambda::Function UPDATE_FAILED Resource update cancelled
|  AppStack appApi/Lambda_GET_--test/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--test1/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--app-list/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_POST_--app/ServiceRole/DefaultPolicy AWS::IAM::Policy UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--test AWS::Lambda::Function UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--test1 AWS::Lambda::Function UPDATE_COMPLETE
|  AppStack appApi/Lambda_GET_--app-list AWS::Lambda::Function UPDATE_COMPLETE
|  AppStack appApi/Lambda_POST_--app AWS::Lambda::Function UPDATE_COMPLETE
|  AppStack FunctionSourcemapUploaderPolicy AWS::IAM::Policy DELETE_COMPLETE
|  AppStack AWS::CloudFormation::Stack UPDATE_ROLLBACK_COMPLETE
⠋  Deploying...

✖  Errors
   AppStack UPDATE_ROLLBACK_COMPLETE
   appApi/Lambda_GET_--app-list: Resource handler returned message: "Uploaded file must be a non-empty zip (Service: Lambda, Status Code: 400, Request ID: 75689c02-22a8-4d51-9cc3-1664bfda0a5e)" (RequestToken: af67eb59-9b37-0c17-8f5d-b95a811ec297, HandlerErrorCode: InvalidRequest)

```

Error logs claim that "Uploaded file must be a non-empty" zip, but I definitely followed `Configuring Java runtime` guides and put the `function.zip` in `build/distributions`, I have no idea where to start to debug this problem.

file structure contained in `function.zip`  like

```
function
├─ com
├─ io
├─ jarkata
├─ lib
├─ META-INF
├─ org
├─ software
└─ application.properties
```

it is **NOT** an empty zip definitely

**Comment by ljd1991** [comment]
After trying, I can reproduce this problem with more specific condition, which is not mentioned before

It seems this problem occurs when define more than 1 handlers in 1 Java projects, and declare more than functions using Java runtime in stacks

e.g.

For Java code, 2 classes are provided like 



Handler1.java

```java
package com.ljd1991;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;

public class Handler1 implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
    @Override
    public APIGatewayProxyResponseEvent handleRequest(APIGatewayProxyRequestEvent input, Context context) {
        System.out.println("hello world!");
        APIGatewayProxyResponseEvent response = new APIGatewayProxyResponseEvent();
        response.setBody("hello world!");
        response.setStatusCode(200);
        return response;
    }
}

```

Handler2.java

```java
package com.ljd1991;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;

public class Handler2 implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
    @Override
    public APIGatewayProxyResponseEvent handleRequest(APIGatewayProxyRequestEvent input, Context context) {
        System.out.println("hello world!");
        APIGatewayProxyResponseEvent response = new APIGatewayProxyResponseEvent();
        response.setBody("hello world!");
        response.setStatusCode(200);
        return response;
    }
}
```



AppStack.ts

```typescript
export function AppStack({stack}: StackContext) {
    const appApi = new Api(stack, "appApi", {
        defaults: {
            function: {
                timeout: 900
            }
        },
        routes: {
            "GET /": {
                function: {
                    runtime: "java11",
                    handler: "com.ljd1991.Handler1::handleRequest"
                }
            },
            "GET /test": {
                function: {
                    runtime: "java11",
                    handler: "com.ljd1991.Handler2::handleRequest"
                }
            }
        }
    })


    // Show the endpoint in the output
    stack.addOutputs({
        AppApiEndPoint: appApi.url
    });
}
```



running `npm run deploy`

error message is like 

```
 appApi/Lambda_GET_--test: Resource handler returned message: "Uploaded file must be a non-empty zip (Service: Lambda, Status Code: 400, Request ID: b1180377-f2d5-4194-9018-cd0f456dc51d)" (RequestToken: 6afc355b-d472-be59-7b98-a61052de1c97, HandlerErrorCode: InvalidRequest)
```

**It  does not happen when declare only one function in stack. I think this is a bug, need to be fixed**

**Comment by jayair** [comment]
Our Java support is mostly community contributed. We could use your help looking into the issue and if possible submitting a PR.

**Comment by fwang** [comment]
Hey @ljd1991, can you try setting `SST_BUILD_CONCURRENCY` to 1, ie.
```bash
SST_BUILD_CONCURRENCY=1 npm run deploy
```
This will make sure the java functions are built one at a time.

Let me know.

**Comment by ljd1991** [comment]
> Hey @ljd1991, can you try setting `SST_BUILD_CONCURRENCY` to 1, ie.
> 
> ```shell
> SST_BUILD_CONCURRENCY=1 npm run deploy
> ```
> 
> This will make sure the java functions are built one at a time.
> 
> Let me know.

Hello @fwang , deployment is successfully completed following your suggestion! It seems concurrency effect Java function deployment. Thanks again!

**Comment by fwang** [comment]
@ljd1991 Thanks for confirming! 谢谢🙏

Fixed in v2.24.23. If you upgrade, u don't need to set `SST_BUILD_CONCURRENCY`
